sharktuner/sharktuner/constraint_generator.py

```diff
@@ -0,0 +1,409 @@
+        assert contraction_dims, "no contraction dimensions"
+        dims = common.ContractionDimensions(
+            batch=list(contraction_dims.batch),
+            m=list(contraction_dims.m),
+            n=list(contraction_dims.n),
+            k=list(contraction_dims.k),
+        )
+
+        res_maps = linalg.get_indexing_maps(root_op)
+        maps = [map_attr.value for map_attr in res_maps]
+        lhs_dims = common.get_map_result_dim_positions(maps[0])
+        rhs_dims = common.get_map_result_dim_positions(maps[1])
+        res_dims = common.get_map_result_dim_positions(maps[2])
+
+        assert lhs_dims, "no lhs dimensions"
+        assert rhs_dims, "no rhs dimensions"
+        assert res_dims, "no result dimensions"
+
+        lhs_type = ir.RankedTensorType(root_op.operands[0].type)
+        rhs_type = ir.RankedTensorType(root_op.operands[1].type)
+        res_type = ir.RankedTensorType(root_op.operands[2].type)
+
+        matmul_size = common.ContractionSizes(
+            M=[lhs_type.shape[lhs_dims.index(dim)] for dim in contraction_dims.m],
+            N=[rhs_type.shape[rhs_dims.index(dim)] for dim in contraction_dims.n],
+            K=[lhs_type.shape[lhs_dims.index(dim)] for dim in contraction_dims.k],
+            B=[lhs_type.shape[lhs_dims.index(dim)] for dim in contraction_dims.batch],
+        )
+
+        self.dims = dims
+        self.matmul_size = matmul_size
+        self.lhs_type = common.ShapedType(lhs_type.shape, lhs_type.element_type)
+        self.rhs_type = common.ShapedType(rhs_type.shape, rhs_type.element_type)
+        self.res_type = common.ShapedType(res_type.shape, res_type.element_type)
+
+    def generate_solutions(
+        self,
+        tuner_context: common.TunerContext,
+        codegen_pipeline: iree_codegen.DispatchLoweringPassPipeline,
+        **pipeline_constraint_options,
+    ) -> Iterator[iree_codegen.CompilationInfoAttr]:
+        num_subgroups = pipeline_constraint_options.get("num_subgroups", 4)
+        mma_intrinsics = pipeline_constraint_options["mma_list"]
+        allowed_waves_per_eu = pipeline_constraint_options.get(
+            "allowed_waves_per_eu", [2]
+        )
+        pipeline_options_search_space = pipeline_constraint_options.get(
+            "pipeline_options_search_space",
+            dispatch_constraints.PipelineOptionsSearchSpace(),
+        )
+        return generate_generic_contraction_solutions(
+            tuner_ctx=tuner_context,
+            contraction_dims=self.dims,
+            matmul_size=self.matmul_size,
+            lhs_type=self.lhs_type,
+            rhs_type=self.rhs_type,
+            res_type=self.res_type,
+            dispatch_kind=common.DispatchKind.contraction,
+            codegen_pipeline=codegen_pipeline,
+            num_subgroups=num_subgroups,
+            mma_intrinsics=mma_intrinsics,
+            allowed_waves_per_eu=allowed_waves_per_eu,
+            pipeline_options_search_space=pipeline_options_search_space,
+        )
```

Comment:
I think the change here and below to just forward `pipeline_constraint_options` got dropped somehow
```suggestion
        return generate_generic_contraction_solutions(
            tuner_ctx=tuner_context,
            contraction_dims=self.dims,
            matmul_size=self.matmul_size,
            lhs_type=self.lhs_type,
            rhs_type=self.rhs_type,
            res_type=self.res_type,
            dispatch_kind=common.DispatchKind.contraction,
            codegen_pipeline=codegen_pipeline,
            **pipeline_constraint_options
        )
```